File: copy_list_with_random.c
#include<stdlib.h>
struct RandomListNode {
 int label;
 struct RandomListNode *next;
 struct RandomListNode *random;
};

void appendNew(struct RandomListNode* head){
    if(!head){
        return;
    }
    struct RandomListNode* current = (struct RandomListNode*)malloc(sizeof(struct RandomListNode));
    struct RandomListNode* head_nxt = head->next;
    current->label = head->label;
    current->next = head_nxt;
    current->random = NULL;
    head->next = current;
    appendNew(head_nxt);
}

void setRandom(struct RandomListNode* head){
    if(!head){
        return;
    }
    if(head->random){
        head->next->random = head->random->next;
    }
    setRandom(head->next->next);
}

void restore(struct RandomListNode* head){
    if(!head){
        return;
    }
    struct RandomListNode* head_nxt = head->next->next, *copy = head->next;
    head->next = head_nxt;
    if(head_nxt){
        copy->next = head_nxt->next;
    }
    else{
        copy->next = NULL;
    }
    restore(head_nxt);
}

struct RandomListNode *copyRandomList(struct RandomListNode *head) {
    if(!head){
        return head;
    }
    appendNew(head);
    setRandom(head);
    struct RandomListNode* ans = head->next;
    restore(head);
    return ans;
}

File: copylistwithrandom.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
# vim:fenc=utf-8
#
# Copyright © 2014 zhao
import sys


class RandomListNode:
  def __init__(self, x):
    self.label = x
    self.next = None
    self.random = None


class Solution:
  # @param head, a RandomListNode
  # @return a RandomListNode
  def copyRandomList(self, head):
    if head is None:
      return None
    mixed = RandomListNode(-1)
    current, prev = head, mixed
    while current:
      new_node = RandomListNode(current.label)
      new_node.random = current.random
      current.random = new_node
      prev.next = new_node
      prev = new_node
      current = current.next
    mixed = mixed.next

    ans = RandomListNode(-1)
    current, prev = mixed, ans
    while current:
      new_node = RandomListNode(current.label)
      cached = current.next
      current.next = new_node
      prev.next = new_node
      prev = new_node
      current = cached
    ans = ans.next

    current = head
    while current:
      current.random.next.random = \
        current.random.random.random.next if current.random.random else None
      current = current.next

    current = head
    while current:
      current.random = current.random.random
      current = current.next
    return ans
